qdrant-landing/content/documentation/tutorials/retrieval-quality.md

```diff
@@ -117,111 +91,126 @@ client.upload_points(  # upload_points is available as of qdrant-client v1.7.1
+
+### Step 4: Measure Retrieval Quality
 
-Qdrant has a built-in exact search mode, which can be used to measure the quality of the search results. In this mode, Qdrant performs a
-full kNN search for each query, without any approximation. It is not suitable for production use with high load, but it is perfect for the 
-evaluation of the ANN algorithm and its parameters. It might be triggered by setting the `exact` parameter to `True` in the search request.
-We are simply going to use all the examples from the test dataset as queries and compare the results of the approximate search with the
-results of the exact search. Let's create a helper function with `k` being a parameter, so we can calculate the `precision@k` for different
-values of `k`.
+To assess retrieval quality, we need to compare the results of an ANN search with the results of an exact search. This comparison will help determine how well the ANN algorithm approximates the true nearest neighbors.
+
+The function `avg_precision_at_k(k)` calculates the average precision at k by comparing the ANN results against the exact search results for each item in the test dataset. Precision at k measures how many of the top k results from ANN match the exact search results.
 
 ```python
 def avg_precision_at_k(k: int):
     precisions = []
     for item in test_dataset:
+        # Perform ANN search
         ann_result = client.query_points(
             collection_name="arxiv-titles-instructorxl-embeddings",
             query=item["vector"],
             limit=k,
         ).points
     
+        # Perform exact search
         knn_result = client.query_points(
             collection_name="arxiv-titles-instructorxl-embeddings",
             query=item["vector"],
             limit=k,
             search_params=models.SearchParams(
-                exact=True,  # Turns on the exact search mode
+                exact=True,  # Enables exact search mode
             ),
         ).points
 
-        # We can calculate the precision@k by comparing the ids of the search results
-        ann_ids = set(item.id for item in ann_result)
-        knn_ids = set(item.id for item in knn_result)
+        # Calculate precision@k
+        ann_ids = set(point.id for point in ann_result)
+        knn_ids = set(point.id for point in knn_result)
         precision = len(ann_ids.intersection(knn_ids)) / k
         precisions.append(precision)
     
     return sum(precisions) / len(precisions)
 ```
 
-Calculating the `precision@5` is as simple as calling the function with the corresponding parameter:
+### Step 5: Measure Retrieval Quality
 
 ```python
-print(f"avg(precision@5) = {avg_precision_at_k(k=5)}")
+print(f"Initial avg(precision@5) = {avg_precision_at_k(k=5)}")
 ```
 
-Response:
+This step measures the initial retrieval quality before any tuning of the HNSW parameters. The HNSW (Hierarchical Navigable Small World) algorithm has two key parameters that influence search performance and quality:
```

Comment:
We could provide a bit more details here:
There are 2 types of parameters which users can tune, index time parameters and search time parameters
index time: `m` and `ef_construct`, search time - `ef`

I think that we might want to mention it here, rather than just add a brief sentence at the end of the article
However, I don't find the code adjustments to be a necessity